### DISCUSSION TITLE
Maplibre Turn-by-Turn Navigation Ideation

### DISCUSSION THREAD

**Discussion by Archdoog** [maplibre/maplibre#210] [Native feature proposals]
# Maplibre Turn-by-Turn Navigation Ideation

## 1. Overview

This discussion outlines ideas and opens up brainstorming for a path forward following the introduction of turn-by-turn navigation to Maplibre. Flitsmeister's forks from Mapbox's early fully open source turn-by-turn navigation SDKs are now available. The maplibre repos for these are https://github.com/maplibre/maplibre-navigation-ios and https://github.com/maplibre/maplibre-navigation-android.

**The main question:**

> What's the path forward for the best outcome when it comes to turn-by-turn navigation?

**This document is a rough draft based on the discussion started in https://osmus.slack.com/archives/C02B2CBSNBU/p1674840496128799. Please @ me to update the content.** Whether it's incorrect information, deletes, edits, or adds. There's a lot covered here and I'm by no means an expert on most of it 😄.

## 2. Current Status

The existing maplibre navigation sdks are forked from the last fully open source Mapbox navigation sdks. They provide a separate fully functioning turn-by-turn navigation system for both iOS and Android. However, they're now quite old and require signficant updates to remain relevant for future developers. Being platform specific, they only support the single platform they were designed for.

## 3. Goals, Considerations & Ideas

1. A turn-by-turn navigation library that provides all basic functionality for driving a route with directions, maneuvers, steps and location updates. Some basics include integration with platform location services, location following, route snapping, reroute detection, route progression, visual instructions, audio instructions and the associated ui providers (e.g. next instruction banner, ETA, lane suggestion).
2. Flexible implementation that allows developers the ability to bring their own route service(s), platform specific UI and more. An example of this could be a public `protocol`/`interface` on iOS/Android that allows a developer to easily implement a route service and decode full instruction routes in designated formats (e.g. OSRM json, PBF, etc). Furthermore, this library could allow flexibility in using soft depencies, service providers and more. E.g. one team may benefit by working with MapboxMaps, MapboxDirections whereas another may want the full DIY maplibre stack. A key concept here may be outlining a handful of existing as well as desired implementations for a specific platforms to define what projects use now and what they want to use.
3. Based on the origins of the existing maplibre libraries, many existing projects may be completely or at least partially using Mapbox SDKs for their mapping needs when they start to consider the transition. It seems that allowing low hurdle support for MapboxMaps, MapboxDirections and any other relevant depenencies would increase the appeal. Furthermore, it's plausible this library could be implemented on top of whatever maps sdk you want, with the exception of automatic route rendering on a sepecific map layer, etc. Though this may not provide the best end result, given the cost of most navigation solutions, it could be worthwhile to increase the appeal of transitioning to maplibre slowly.
4. Focus on latest generation SDKs and UI for the primary supported platforms. In the context of iOS this means easy implementation on SwiftUI (with secondary support for UIKit as needed). On Android this would mean easy implementation with Jetpack Compose (with secondary support for Fragment/Activity). This same concept of forward thinking can be applied to many more platforms.
5. What else?

## 4. General Questions

The following questions are generally applicable to most of the plans discussed below.

1. What are the target use cases and/or industries? (e.g. native mobile, web/PWA?, embedded, all the way up to things like fleet and ADAS?).
2. What are the core platforms to support with an easy to use wrapper? (e.g. Swift iOS, Kotlin Android, Android Auto, Flutter, Qt, etc, etc).
3. What route services should be the priority for ease of integration to provide full turn-by-turn instructions with route data? 
4. What route formats should be supported?
5. Should commonly used platforms get default UI implemenations?
6. Should the library require all or nothing maplibre adoption? Or should it be flexible and used with many bends of mapping, routing, etc?
7. What are the highest priority languages for maplibre as a whole? This could increase cross project collaboration.
8. What languages and toolchains are the best for growth with newer programmers & the future?
9. How can the library leverage latest generation technologies to provide an improved navigation experience? What outside of the box ideas exist for this problem.

## 5. Prospective Plans

The following outlines some ideas with potential context.

### A. General Purpose Multi-platform (e.g. Rust, C++, etc.)

Use a general purpose language to build the core navigation SDK. Provide c bindings and native language wrappers for supported platforms. This method is similar to the architecture of https://github.com/maplibre/maplibre-gl-native (maplibre-gl-native is C++). 

**Benefits**

1. **Fast, stable and flexible** - The library will run natively on many platforms and architectures. It can offer c bindings for easy integration into many platform stacks.
2. **Future proofing** - Since the library can be built natively on most platforms, it should remain applicable to new platforms without major changes in direction. 
3. **Platform agnostic** - A general purpose library will allow developers outside of the maplibre ecosystem and supported platforms to implement their own use case. E.g. an automotive multimedia system that runs a proprietary UI constructed with Qt, etc.
4. **Support for regulated platforms?** - Building a library with a lower level language may open the door for system critical functionality. An example could be safety regulated linux that runs safety criticial automotive systems. This is not necessarily a target for support, but the door could remain open with this plan.
5. **Core libraries** - Access to a wide spectrum of general purpose libraries (e.g. https://github.com/valhalla/valhalla, https://georust.org and many more). Some of which can be very challenging to implement on a per platform basis with platform languages and SDKs like Swift/Kotlin. While this may not be of major relevance to this SDK specifically, it does open the door for developers to integrate this library into a larger blend of projects.
6. **Flexible build systems** - This can also be a challenge, but these languages typically allow the most amount of flexiblity when it comes to building for many platforms and archs (e.g. using `CMAKE` or `cargo`). 

**Drawbacks or Challenges**

7. **Platform specific code** - Supported platforms will take additional work to build platform specific wrappers. Depending on the tools available and complexity, this can be quite a challenge. An example is a Swift wrapper library that enteracts with the core library's c bindings to provide iOS developers logical and easy to use implementation tools in Swift. 
8. **Platform specific package managers** - Package manager support can be quite challenging on some platforms and many will be required.
9. **Large scope of developer skills** - Requires lots of specialities to build and maintain.

**Questions**

- What language? There are many costs and benefits. C++ has decades of dominance but more modern alternatives exist to solve the challenges C++ presents. https://survey.stackoverflow.co/2022/#technology-most-loved-dreaded-and-wanted. 

### B. Limited Multi-Platform (e.g. Kotlin MultiPlatform)

A mutliplatform framework designed for a specific blend of platforms. I haven't explored this much yet, so feel free to add points to it.

**Benefits**

1. **Easy integration** - More easily integrated with some platforms.
2. **Less scope of developer skills** - Since the language is more closely coupled with certain platforms, it's likely to require a smaller scope of specialities.

**Drawbacks or Challenges**

3. **Support limitations** - It's likely roadblocks will exist outside of the core supported platforms. This could limit project applicability and growth.
4. **Major growth** - Many of these libraries are newer and changing quite rapidly, this could present a challenge for long term maintenance.

**Questions**

- What framework is the best?

### C. Separate per Platform (Existing)

This represents the existing `maplibre-navigation-ios` and `maplibre-navigation-android` where each platform has a separate SDK developed directly in its native language and with native libraries and SDKs.

**Benefits**

1. **Direct integration** - A platform specific library written in the platform's native language makes it extremely easy to use on that platform.
2. **Accessible for platform developers** - The library is likely to be easier to work on for developers of each platform. This may encourage growth on that specific platform.
3. **Quickly leverage new platform features** - Since the library will be written in the platform's language and directly support all platform SDKs it should be very easy to support new platform features.

**Drawbacks or Challenges**

4. **Duplication** - Each platform will require a duplicated core library that provides the same functionatily. Any new platform has to re-emplement that entire code base.
5. **Limited use cases** - Since each platform requires it's own dedicated library, it's likely the use case scope will be limited through less incentive to collaborate.
6. **Many repos, pipelines and maintainers** - This approach means each platform has it's own repo, it's own CI & CD, and it's own maintainer. 

### D. Other?

What other plans exist to accomplish the goals?

## 6. References 

1. **Slack Discussion #1** - https://osmus.slack.com/archives/C02B2CBSNBU/p1674840496128799
2. **Forked Mapbox Navigation iOS** - https://github.com/maplibre/maplibre-navigation-ios
3. **Forked Mapbox Navigation Android** - https://github.com/maplibre/maplibre-navigation-android
4. **Latest Developer Preferences and Skillsets** - https://survey.stackoverflow.co/2022/
5. **Rust on iOS, Android & Flutter** https://github.com/robertohuertasm/rust-for-android-ios-flutter
6. **Kotlin Multiplatform**
7. Many more that weren't mentioned in the initial Slack discussion...

# Outcome

TBA based on discussion...

**Comment by wipfli** [comment]
Would it also be able to do public transit instructions?

Like for example here: https://support.apple.com/guide/iphone/get-transit-directions-ipha44f57caa/ios

<img src="https://user-images.githubusercontent.com/53421382/215291194-3f83093d-bdae-4388-8690-10742257081f.png" width=250>

**Reply by boldtrn** [reply]
I guess this would be possible, but would need some customization. You don't really have cases like rerouting etc.

I think for PT, it might be easier to use something like geofencing, if you get closer than X meter to a station, notify the user. If you are at the station, notify the user.

**Comment by boldtrn** [comment]
This is an interesting discussion, with a lot of different topics:

- UI: The navigation library should work without any specific UI, but should allow devs to build their own UI. A sample UI is definitely great, but I think we can't do this cross platform.
- Cross Platform: I am leaning towards Kotlin Multi Platform. Because I can see a path where we can do this incrementally. Unless we have funding, I doubt we can do a full rewrite.
- I think the concept Mapbox chose with the navigation library makes sense. To have a common core, for the core features, and implement platform specifics per platform.

**Reply by Archdoog** [reply]
- UI: Definitely agree here. I think it makes sense to keep the shared functionality simple with the intent that any system and its UI can be implemented on top if there's a will to do it. 
- Cross Platform: I think it would be useful to get some more info on benefits and challenges of kotlin multi platform. E.g. build system, 
- Agree here as well. This would also keep the door open for future platforms, use cases, etc.

**Comment by louwers** [comment]
Adding this valuable Slack comment from @1ec5 here so it doesn't get lost. :slightly_smiling_face: 

> I haven’t looked closely at this fork, but fwiw the iOS repo seems to be based on a gratuitously old version of Mapbox’s navigation SDK. v0.21 was indeed the last version before the closed-source navigation-native dependency was added, but even in v2.x it isn’t functionally required. v0.22 just moved the open-source implementation to LegacyRouteController, but there’s still a public option to switch to that implementation, all the way up to [the last open-source version](https://github.com/mapbox/mapbox-navigation-ios/issues/3115) and beyond.

> > Years ago, someone contributed a PR to add support for the raw OSRM response format, which is the basis for the Mapbox Directions API v5 format.
>
> https://github.com/mapbox/mapbox-directions-swift/pull/198

**Reply by Archdoog** [reply]
Per this it seems like a useful starting point could be updating iOS and android to these last ISC/MIT versions (or creating some new parallel projects for some experimentation). We could start by removing the mapbox private binary then start working out the process of creating a new shared navigation core that plugs into these. This could be done while supporting any remaining legacy functionality like the `LegacyRouteController` in the  iOS repo.

That process seems like a reasonable plan to keep things manageable while also maintaining a more complete experience for iOS and Android. Once that job is done, it would then be natural to start adding support for other frameworks, platforms, etc.

**Reply by boldtrn** [reply]
> We could start by removing the mapbox private binary then start working out the process of creating a new shared navigation core that plugs into these.

I had the same idea some time ago. I wasn't able to get it working. It's not impossible, but far from a quick fix.

**Comment by ianthetechie** [comment]
Thanks for this writeup!

I think the goals are spot on. A few general thoughts from me:

I think the approach MapLibre GL Native took works fairly well. It's not perfect by any means, but a shared common codebase with platform-specific frontends works well. This lets each platform leverage the latest tools. These change every few years (ex: XIB files + Obj-C -> Storyboards + Swift -> SwiftUI), and I am personally somewhat skeptical that tools like Kotlin Multiplatform will keep up or gain enough traction.

As such, I'd favor using some common language. It could be C++ or Rust; I personally prefer the latter, but that's just one man's opinion. The core reason for my preference (aside from all the usual reasons people like Rust incl. memory safety) is that Rust binding generators are generally quite advanced, in part because Rust supports a C ABI for its FFI. Obviously I'm not going to suggest writing this in raw C, but I personally feel the C++ choice has the drawback of not having the simplicity of C or binding generators that are as... polished(?). It's _possible_ to integrate C++ code with an iOS app using the Obj-C++ layer, and Rust similarly has CXX support via several crates, but it's _always_ a chore. And that's on a platform with clang + llvm-based tooling. The story is surely much worse on Android. I'd suggest looking at something like this (proposed by Patrick Fitzgerald in Slack): https://mozilla.github.io/uniffi-rs/.

> What are the core platforms to support with an easy to use wrapper? (e.g. Swift iOS, Kotlin Android, Android Auto, Flutter, Qt, etc, etc).

We can start with Swift iOS and Kotlin Android. I think CarPlay is supported in the current forks btw, but I'm not totally sure on status. I also think Flutter is going to be pretty high on the list growth-wise, but for now I think most are just using native SDKs since that's all Mapbox (and others?) ever offered. (This also answers question 4.7 and 4.8.)

Jumping ahead a bit, I think we can separate maintainership for the individual platforms and the core to get the best people working on each. If the core is solid, then the various "frontends" can be a bit more free to choose the best UI tooling etc. Ideally the core API will have enough callbacks/hooks so that the "business logic" duplication in the frontends is kept to a minimum. Unfortunately I'm probably not the best qualified to spec out what all of these are, but will contribute what I can :)

> Should commonly used platforms get default UI implemenations (sic)?

Yes.

> Should the library require all or nothing maplibre adoption? Or should it be flexible and used with many bends of mapping, routing, etc?

I'm not quite sure what you mean. Yes, it should require the MapLibre map SDK, but that's already cross-provider. MapLiber is an organization focusing on the map display/visualization problems. While there are orgs represented who also specialize in routing (such as mine, Stadia Maps), there is no MapLibre navigation standard, and I rather doubt there will be.

I would say for routing data, this should be pluggable using interfaces/protocols. This will have both a request and response halves, and these should not be tightly coupled. For example, the Mapbox Directions v5 response format is basically OSRM (plus some stuff?), and the Valhalla engine (hosted by many providers, including us) can support both OSRM as well as its own JSON format and protobuf. On the request side, while Valhalla supports an OSRM output, the request API is completely different. So being able to write a Valhalla-specific request model but use a common OSRM response model will bring a ton of flexibility and reduce work.

**Comment by Jab2870** [comment]
Hi,

Just wanted to chime in and ask if this is ever likely to make it's way to maplibre js. Skimming over the previous discussion, there is a lot of talk about native ios and android integration. I'd love to be able to use this in something like a PWA or for a rpi based sat nav using the JS apis though if it were possible.

**Reply by HarelM** [reply]
I think that for gl js you can either write a plugin or use the following plugin: 
https://github.com/maplibre/maplibre-gl-directions

**Reply by Jab2870** [reply]
thanks for the super quick response.

Thanks for the link. However, unless I've simply missed it, I thought the gl-directions plugin was for showing a route overview, rather than for showing the current position and next steps.

![](https://github.com/maplibre/maplibre-navigation-android/raw/main/.github/preview.png)

The screenshot from the android version looks as though it's designed to be mounted in the car whilst your driving, updating your location as you drive. Unless I'm mistaken, the gl-directions dosen't do that.

Apoologies in advance if I've just missed it.

**Reply by HarelM** [reply]
I don't know, you can check with the plugin author or simply create a plugin with your desired functionality. 😃

**Reply by 1ec5** [reply]
Sounds like what https://ffwdmejs.org/ was pursuing. That project got started at around the same time as the Mapbox Navigation SDK (the origin of what’s now the MapLibre Navigation SDK) but focused on the Web instead of native platforms.

**Reply by boldtrn** [reply]
You can have a look at [GraphHopper Maps](https://github.com/graphhopper/graphhopper-maps/pull/279) as well - you can try this [here](https://navi.graphhopper.org).

**Reply by ianthetechie** [reply]
We're working on something similar as well for [Ferrostar](https://github.com/stadiamaps/ferrostar). Here is the [web proof of concept](https://stadiamaps.github.io/ferrostar/web-demo/).

**Comment by ivannovazzi** [comment]
One challenge when developing navigation SDKs is having realistic, controllable vehicle data to test against — especially for edge cases like rerouting or turn instruction timing.

[Moveet](https://github.com/ivannovazzi/moveet) might help here: it's an open-source fleet simulator that drives synthetic vehicles on real road networks (OpenStreetMap/GeoJSON) using A* pathfinding, streaming live positions over WebSocket. You get realistic GPS coordinates that follow actual road geometry, which you can feed into a navigation SDK under development to test how it responds to real-world movement patterns.

Not directly a navigation SDK itself, but a useful tool for generating the vehicle movement data that navigation testing requires.
